### PR TITLE
New Ruff Rules

### PR DESCRIPTION
**Pull request by tiniuspre**
# Added new ruff rules

## [Rules](https://docs.astral.sh/ruff/rules/)
- [**S** =  Flake8 Bandit](https://docs.astral.sh/ruff/rules/#flake8-bandit-s)   
- [**DJ** = Flake8 Django](https://docs.astral.sh/ruff/rules/#flake8-django-dj)
- [**C90** = McCabe](https://docs.astral.sh/ruff/rules/#mccabe-c90)
- [**W** = Warnings](https://docs.astral.sh/ruff/rules/#warning-w)  
- [**N** = Pip8 Naming](https://docs.astral.sh/ruff/rules/#pep8-naming-n)   
- [**UP** = Pyupgrade](https://docs.astral.sh/ruff/rules/#pyupgrade-up)  
- [**FBT** = Flake8 boolean trap](https://docs.astral.sh/ruff/rules/#flake8-bandit-s) 
- [**B** = Flake8 Bugbear](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b)   
- [**C4** = Flake8 comprehensions](https://docs.astral.sh/ruff/rules/#flake8-comprehensions-c4) 
- [**T20** = Flake8 print](https://docs.astral.sh/ruff/rules/#flake8-print-t20) 
- [**PYI** = Flake8 pyi](https://docs.astral.sh/ruff/rules/#flake8-pyi-pyi) 
- [**RET** = Flakke8 return](https://docs.astral.sh/ruff/rules/#flake8-return-ret)
- [**SIM** = Flake8 Simplify](https://docs.astral.sh/ruff/rules/#flake8-simplify-sim)
- [**TID** = Flake8 Tidy imports](https://docs.astral.sh/ruff/rules/#flake8-tidy-imports-tid)
- [**PERF** = Perflint](https://docs.astral.sh/ruff/rules/#perflint-perf)

## Options
C90 limits the [complexity](https://docs.astral.sh/ruff/rules/complex-structure) of functions. It is currently sett to 5.

> The McCabe complexity of a function is a measure of the complexity of the control flow graph of the function. It is calculated by adding one to the number of decision points in the function. A decision point is a place in the code where the program has a choice of two or more paths to follow.
Functions with a high complexity are hard to understand and maintain.

Functions already in the code with high complexity > 5 have been ignored.

Other complex and time consuming changes have also been ignored. 

## Notes
Disabled [DJ001](https://docs.astral.sh/ruff/rules/django-nullable-model-string-field/)

> Checks nullable string-based fields (like CharField and TextField) in Django models.

> If a string-based field is nullable, then your model will have two possible representations for "no data": None and the empty string. This can lead to confusion, as clients of the API have to check for both None and the empty string when trying to determine if the field has data.
The Django convention is to use the empty string in lieu of None for string-based fields.

Allot of the char/text fields have null set to true, unsure if things might fuck up in frontend if this is removed.

### CODE REVIEW

File: backend/pyproject.toml [hunk 1]
```diff
@@ -33,6 +35,8 @@ preview = true
 explicit-preview-rules = true
 ignore-init-module-imports = true # This implicit export strategy is used in this project.
 
+dummy-variable-rgx = '^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$'
```
Comment (emilte) [inline backend/pyproject.toml:38]:
```suggestion
dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
```

File: backend/pyproject.toml [hunk 2]
```diff
@@ -61,6 +65,22 @@ select = [
...
     "D211", # pydocstyle
     "T20", # flake8-print
+    "S", # https://docs.astral.sh/ruff/rules/#flake8-bandit-s
+    "DJ", # Flake8-Django
+    "C90", # maccabe
+    "W", # pycodestyle (Warning)
+    "N", # flake8-broken-line
+    "UP",
+    "FBT",
+    "B",
```
Comment (emilte) [inline backend/pyproject.toml:75]:
```suggestion
```

File: backend/root/custom_classes/permission_classes.py [hunk 3]
```diff
@@ -45,7 +45,6 @@ def has_permission(self, request: Request, view: APIView) -> bool:  # noqa: PLR0
 
         if request.method == 'POST':
             post_perms: list[str] = self.get_required_permissions(method='POST', model_cls=model_cls)
-            has_perm = user.has_perms(perm_list=post_perms)
-            return has_perm
+            return user.has_perms(perm_list=post_perms)
```
Comment (emilte) [inline backend/root/custom_classes/permission_classes.py:48]: Hvilken regel er dette?
Her foretrekker jeg egt forklarende variabelnavn. Det er også enklere å sette breakpoints.

File: backend/root/management/commands/generate_routes.py [hunk 4]
```diff
@@ -54,8 +54,7 @@ def parse_name(name: str) -> str:
     Parse django url name to frontend route format.
     'scope:some-name' -> 'scope__some_name'
     """
-    scoped_snake_name = name.replace(':', '__').replace('-', '_').replace('/', '')
-    return scoped_snake_name
+    return name.replace(':', '__').replace('-', '_').replace('/', '')
```
Comment (emilte) [inline backend/root/management/commands/generate_routes.py:57]: Dette reduserer lesbarhet imo

File: backend/root/management/commands/generate_routes.py [hunk 5]
```diff
@@ -109,11 +108,9 @@ def parse_url(url: str) -> str:
         sub_paths[i] = new_sub_path
 
     # from: ['', 'some', 'path', ':orgId', '']
-    new_url = '/'.join(sub_paths)
+    return '/'.join(sub_paths)
     # to: '/some/path/:orgId/'
 
-    return new_url
-
```
Comment (emilte) [inline backend/root/management/commands/generate_routes.py:116]: .

File: backend/root/management/commands/seed.py [hunk 6]
```diff
@@ -22,7 +21,7 @@ def print_progress(
     progress: int,
     prefix: str = '',
     suffix: str = '',
-    start_time: Optional[float] = None,
+    start_time: float | None = None,
```
Comment (emilte) [inline backend/root/management/commands/seed.py:24]: I mine øyne signaliserer ikke dette det nøyaktig det samme.
Hvilken regel er dette?

File: backend/root/management/commands/seed_scripts/samf3.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 import csv
 import os.path
 import multiprocessing
-from typing import Iterator
+from collections.abc import Iterator
```
Comment (emilte) [inline backend/root/management/commands/seed_scripts/samf3.py:6]: Forskjellen er?

File: backend/root/custom_classes/permission_classes.py (hunk 3 above)
Comment (tiniuspre) [reply backend/root/custom_classes/permission_classes.py:48]: https://docs.astral.sh/ruff/rules/unnecessary-assign/

Syntes det tydeliggjør hva man returnerer. 

Ser ikke problemet med breakpoints det blir jo i grunn det samme

File: backend/root/management/commands/seed.py (hunk 6 above)
Comment (tiniuspre) [reply backend/root/management/commands/seed.py:24]: https://docs.astral.sh/ruff/rules/non-pep604-annotation/

File: backend/root/management/commands/seed_scripts/samf3.py (hunk 7 above)
Comment (tiniuspre) [reply backend/root/management/commands/seed_scripts/samf3.py:6]: https://docs.astral.sh/ruff/rules/deprecated-import/

File: backend/pyproject.toml (hunk 1 above)
Comment (emilte) [reply backend/pyproject.toml]: Jeg ønsker å resolve mine egne tråder

File: backend/root/custom_classes/permission_classes.py (hunk 3 above)
Comment (emilte) [reply backend/root/custom_classes/permission_classes.py:48]: Jeg synes den er litt for streng. 
Jeg ønsker ofte å ha variabelen tilgjengelig under debugging, dene regelen ødelegger litt

File: backend/root/management/commands/seed.py (hunk 6 above)
Comment (emilte) [reply backend/root/management/commands/seed.py:24]: Litt uheldig at regelen tar for seg to ulike ting.
Jeg liker ikke at den ødelegger Optional, men det er verdt det for å få den andre

File: backend/root/custom_classes/permission_classes.py (hunk 3 above)
Comment (emilte) [reply backend/root/custom_classes/permission_classes.py:48]: Å tydeliggjøre gjør den ikke imo. Her grøsser jeg meg for lange oneliners uten forklarende navn

File: backend/samfundet/serializers.py [hunk 8]
```diff
@@ -349,22 +349,17 @@ def _permission_to_str(permission: Permission) -> str:
...
         # Collect user-level and group-level object permissions.
         user_object_perms_qs = UserObjectPermission.objects.filter(user=user)
         group_object_perms_qs = GroupObjectPermission.objects.filter(group__in=user.groups.all())
 
-        perm_objs = []
-        for obj_perm in itertools.chain(user_object_perms_qs, group_object_perms_qs):
-            perm_objs.append(self._obj_permission_to_obj(obj_perm=obj_perm))
-
-        return perm_objs
+        return [self._obj_permission_to_obj(obj_perm=obj_perm) for obj_perm in itertools.chain(user_object_perms_qs, group_object_perms_qs)]
```
Comment (emilte) [inline backend/samfundet/serializers.py:362]: .

File: backend/samfundet/views.py [hunk 9]
```diff
@@ -435,7 +435,7 @@ class AllGroupsView(ListAPIView):
 
 @method_decorator(ensure_csrf_cookie, 'dispatch')
 class CsrfView(APIView):
-    permission_classes: list[Type[BasePermission]] = [AllowAny]
+    permission_classes: list[type[BasePermission]] = [AllowAny]
```
Comment (emilte) [inline backend/samfundet/views.py:438]: .

File: backend/samfundet/views.py [hunk 10]
```diff
@@ -627,10 +625,7 @@ def get_queryset(self) -> QuerySet[User]:
             default=None,
             output_field=None,
         )
-        users_without_interviews = (
-            User.objects.filter(admissions__recruitment=recruitment).annotate(num_interviews=Count(interview_times_for_recruitment)).filter(num_interviews=0)
-        )
-        return users_without_interviews
+        return User.objects.filter(admissions__recruitment=recruitment).annotate(num_interviews=Count(interview_times_for_recruitment)).filter(num_interviews=0)
```
Comment (emilte) [inline backend/samfundet/views.py:628]: Dette er langt fra enklere

Comment (tiniuspre) [reply backend/samfundet/views.py:628]: Skal fjerne regelen

File: backend/samfundet/serializers.py (hunk 8 above)
Comment (tiniuspre) [reply backend/samfundet/serializers.py:362]: Klager på 2 regler
1: https://docs.astral.sh/ruff/rules/manual-list-comprehension/ 
2: Return regelen

File: backend/samfundet/views.py (hunk 9 above)
Comment (tiniuspre) [reply backend/samfundet/views.py:438]: https://docs.astral.sh/ruff/rules/non-pep585-annotation/

Comment (emilte) [reply backend/samfundet/views.py:438]: Var ikke klar over at type og Type var likt.

File: backend/samfundet/serializers.py (hunk 8 above)
Comment (emilte) [reply backend/samfundet/serializers.py:362]: comprehension er ikke alltid så lesbart. Kan ignoreres her

File: backend/samfundet/views.py (hunk 10 above)
Comment (tiniuspre) [reply backend/samfundet/views.py]: Fjernet regelen

File: backend/samfundet/serializers.py (hunk 8 above)
Comment (tiniuspre) [reply backend/samfundet/serializers.py]: Done

File: backend/root/management/commands/generate_routes.py (hunk 5 above)
Comment (tiniuspre) [reply backend/root/management/commands/generate_routes.py]: Done

File: backend/root/custom_classes/permission_classes.py (hunk 3 above)
Comment (tiniuspre) [reply backend/root/custom_classes/permission_classes.py]: Endret

File: backend/pyproject.toml
Comment (emilte) [inline backend/pyproject.toml:1]: Til neste gang er det enklere å dele opp i mange små PR enn én stor, det blir mange filer å se over :) 

fyi så har jeg sluttet å godkjenne, men jeg fortsetter å se over. Dette er for å få i gang dere aktive webbere ^^

Comment (emilte) [reply backend/pyproject.toml:1]: Da kunne du fått inn et par regler underveis mye kjappere

File: backend/root/utils/mixins.py [hunk 11]
```diff
@@ -57,6 +57,44 @@ def save(self, *args, **kwargs):
     class Meta:
         abstract = True
 
+    def save(self, *args: Any, **kwargs: Any) -> None:
```
Comment (emilte) [inline backend/root/utils/mixins.py:60]: Hva skjer med disse?

Comment (tiniuspre) [reply backend/root/utils/mixins.py:60]: Ikke min endring, ser bare sånn ut siden jeg flyttet på funksjonen

Comment (tiniuspre) [reply backend/root/utils/mixins.py:60]: https://docs.astral.sh/ruff/rules/django-unordered-body-content-in-model/

**Review by Snorre98 — APPROVED** [review]
LGTM

**Pull request merged**